### PR TITLE
Add warning for 1-step training if synapse=None

### PR DESCRIPTION
**Pull request by drasmuss**
Synapses introduce a one-timestep delay, so it is almost certainly a mistake to be training a network for one timestep with non-`None` synapses.  This adds a warning to help users diagnose this issue (otherwise users just end up with poor training performance, which is hard to figure out).

Addresses #54

### CODE REVIEW

**Comment by hunse** [comment]
That's fair. I hadn't thought of that false positive, either. Of course, it would theoretically be possible to check for non-null synapses in the loss function path only, but that's probably more than we need to do right now. (Unless there's some smart way to get TensorFlow's autodiff to catch this for us. Have an error if it ever tries to differentiate a synapse over only one timestep. Though how does it know where a synapse is? It would have to look for recurrence. Could it be as simple as finding loops in the op graph and erroring if we've only got one timestep?)

I like the idea of reducing the output. Here's what I currently get for a basic MNIST model:
```
Build finished in 0:00:01                                                                                           
Optimization finished in 0:00:01                                                                                    
Construction finished in 0:00:01                                                                                    
2018-09-05 14:20:19.317004: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-09-05 14:20:19.735390: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:895] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-09-05 14:20:19.735682: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1105] Found device 0 with properties: 
name: GeForce GTX 970M major: 5 minor: 2 memoryClockRate(GHz): 1.038
pciBusID: 0000:01:00.0
totalMemory: 5.94GiB freeMemory: 5.88GiB
2018-09-05 14:20:19.735707: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1195] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 970M, pci bus id: 0000:01:00.0, compute capability: 5.2)
error before training: 89.90%
/data/eric/workspace/nengo_dl/nengo_dl/simulator.py:508: UserWarning: Training for one timestep, but the network contains synaptic filters (which will introduce at least a one-timestep delay); did you mean to set synapse=None?
```
I don't care about the CPU instruction support, particularly if I'm not running on CPU (duh, TensorFlow!). I have no idea what that NUMA thing is. I don't need to know that it found my GPU, and the PCI bus info and stuff. I do like knowing what device it's running on, i.e. that it's actually using my GPU instead of my CPU. So that "Creating TensorFlow device" line is useful. I have no idea how possible it is to suppress some of these lines and not others.

**Comment by drasmuss** [comment]
Unfortunately there's no way to selectively disable parts of that, it's just all or nothing.  We could disable it, and then add in a log message on the NengoDL side saying what device you're running on (I agree that that's useful to know)?

**Comment by drasmuss** [comment]
Also worth noting that us disabling that TensorFlow log output depends on NengoDL being imported before TensorFlow.  The TensorFlow log level is set on import, so if someone does
```
import tensorflow
import nengo_dl
```
there's nothing we can do about it.

**Comment by drasmuss** [comment]
Actually was just testing that out more carefully, and I was wrong.  It looks like the TensorFlow log level is set when you create the first `Session`, not on import.  So that's good news, means that we should be able to pretty reliably disable the logging if we want.

**Comment by drasmuss** [comment]
Added a commit that disables the TensorFlow logging and adds in a log message on the NengoDL side indicating which device(s) the simulation will be running on.

File: nengo_dl/simulator.py [hunk 1]
```diff
@@ -117,15 +117,16 @@ def __init__(self, network, dt=0.001, seed=None, model=None,
...
-                    "install tensorflow-gpu (`pip install tensorflow-gpu`).")
+        installed_dists = [d.project_name for d in pkg_resources.working_set]
+        if device is None and ("tensorflow-gpu" not in installed_dists and
+                               "tf-nightly-gpu" not in installed_dists):
+            warnings.warn(
+                "No GPU support detected. It is recommended that you "
+                "install tensorflow-gpu (`pip install tensorflow-gpu`).")
+            logger.info("Running on CPU")
+        else:
+            logger.info("Running on %s", "CPU/GPU" if device is None else (
```
Comment (hunse) [inline nengo_dl/simulator.py]: I would maybe put "Unknown device" or something if `device is None`. "CPU/GPU" makes it sound like it's using both the CPU and GPU.

Comment (drasmuss) [reply nengo_dl/simulator.py]: That is what happens (if you don't explicitly specify a device TensorFlow will assign ops to devices according to internal rules, which will likely involve a mix of CPU and GPU execution).

Comment (hunse) [reply nengo_dl/simulator.py]: Ahh, I see. I just want to be sure that the GPU is definitely being used in that case. Is it possible for `tensorflow-gpu` to be installed, but e.g. CUDA not be set up right so the GPU doesn't actually get used?

Comment (drasmuss) [reply nengo_dl/simulator.py]: If you have `tensorflow-gpu` installed then it will try to load the CUDA libraries on import, and you'll get an error if it can't find them (just double checked that that is true).  Same if you have an incompatible version of CUDA installed.  So I'm pretty sure you can be confident that if you have `tensorflow-gpu` installed and you aren't seeing any errors, it's using the GPU.  

It's possible that there could be some obscure way to break it past that?  But that'd be a situation where TensorFlow believes it is running on the GPU, but it isn't actually running on the GPU.  TensorFlow doesn't silently defer to the CPU in case of errors, it just throws the error it seems.

As a sidenote, if you'd like to see which ops are assigned to which device, you can use

```
nengo_dl.configure_settings(session_config={"log_device_placement": True})
```

Also note that if anyone ever wants these log messages back they can be re-enabled via
```
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
```

Comment (drasmuss) [reply nengo_dl/simulator.py]: A slight amendment to that.  I think it might be possible to have an error where CUDA is installed but it isn't detecting your GPU.  In this case TensorFlow will assign all the ops to the only available device (CPU).  However, if you explicitly assign ops to a GPU device, you will get an error if none is detected.  So I should revise the statement above to:

> You can be confident that if you aren't seeing any errors when running with `device="/gpu:0"`, it's using the GPU.

And then you can also be sure that if you can run models with `device="/gpu:0"`, `device=None` can also run on the GPU (although it may not, depending on TensorFlow's internal placement logic).
